### ISSUE TITLE
Troubles with HttpClient

### ISSUE THREAD

**Issue by fatum** [ratpack/ratpack#1168]
We use Ratpack on highload project which process up to 30K RPS. Main goal of project is responding to [RTB](https://en.wikipedia.org/wiki/Real-time_bidding) auctions and making some decisions to show most relevant advertising for user. Ratpack is good fit for this and works wonderfully well on 4-8 dynamically scaled spot instances on AWS (c4.2xlarge) with JRuby 9.1.2.

Decision logic is CPU bound - parsing 2KB json, matching advertising campaigns, preparing output (building XML, json) etc.

But, when we started to make network requests to our advertiser's backends on some fraction of auctions (1-2K RPS total) we encountered on some problems:

1. Without any network IO CPU utilization is below 20-25%
2. With http requests CPU utilization is increasing between 5-10% and doing well for 3-4 hours
3. After 3-4 hours there are very strange CPU spikes not related (probably) to GC (judging by VisualVM). Btw, Old Gen grows very quickly but cleared without any slowdowns or CPU spikes (screenshots from VisualVM is attched)
4. These spikes become more frequent and long and ELB marks host as unhealthy
5. The trouble stays when we test bigger hosts (16 CPU, 32GB RAM), try CMS, G1 algorithms with different settings

### How we make http requests

```ruby
timer = create_timer 'networks.request'

# Network class creates http request and wraps its promise using Hystrix's Command
promises = networks.map { |n| NetworkRequest.new(n, request).execute }

ParallelBatch.of(promises).yieldAll.flatMap do |results|
  timer.stop # Not executed without Hystrix's when request failed on timeout

  # Process results
end
```
http request

```ruby
client = HttpClient.of do |spec|
   spec.poolSize(50)
   spec.readTimeout(Java.java.time.Duration.ofMillis(NETWORK_TIMEOUT)) # Wait response only 300ms
end

client.post(uri) { |req| 
  req.body { |req_body| req_body.text(body) } 
}.flatMapError { |e|
  # Fail for Hystrix
  Promise.error Error::MediationFail.new("[#{network[:name]}] #{e.to_s}", state)
}.flatMap { |response|
  # Prepare response and return Promise.value(data)
}
```

Hystrix command

```ruby
class Command < HystrixObservableCommand
      def initialize(request, params = {})
        @request = request

        command_name = params.fetch(:name, 'mediation_command').to_s
        circuit_breaker = params.fetch(:circuit_breaker, true)
        threshold = params.fetch(:threshold, 2).to_i
        timeout = params.fetch(:timeout, 300).to_i

        setter = Setter.
          withGroupKey(HystrixCommandGroupKey::Factory.asKey("Mediation")).
          andCommandKey(HystrixCommandKey::Factory.asKey(command_name)).
          andCommandPropertiesDefaults(
            HystrixCommandProperties.Setter().
              withExecutionIsolationStrategy(HystrixCommandProperties::ExecutionIsolationStrategy::SEMAPHORE).
              withCircuitBreakerSleepWindowInMilliseconds(1000).tap do |config|
                config.withCircuitBreakerEnabled(circuit_breaker)
                if circuit_breaker
                  config.withCircuitBreakerRequestVolumeThreshold(threshold)
                end

                config.withExecutionTimeoutEnabled(!timeout.zero?)
                unless timeout.zero?
                  config.withExecutionTimeoutInMilliseconds(timeout)
                end
              end
          )

        super setter
      end

      def construct
          RxRatpack.observe @request
      end

      def resumeWithFallback
        RxRatpack.observe Promise.value(nil)
      end

      def self.promise(request, params = {})
        return Promise.value(nil) if request.blank?

        RxRatpack.promise self.new(request, params).toObservable
      end
    end
```

### Screenshots

![screen shot 2017-02-21 at 14 21 19](https://cloud.githubusercontent.com/assets/313161/23330460/4493ff58-fb5f-11e6-8b63-381c2e9b9b9c.png)
![screen shot 2017-02-21 at 14 40 12](https://cloud.githubusercontent.com/assets/313161/23330463/4933dff6-fb5f-11e6-8da0-f96757893a07.png)
![screen shot 2017-02-25 at 12 56 27](https://cloud.githubusercontent.com/assets/313161/23330464/4adc4442-fb5f-11e6-8cd1-174b3a6f394d.png)

### Our stack

OpenJDK 8
Ratpack 1.4.5
JRuby 9.1.2.0 (thanks to @headius)

**Comment by fatum** [comment]
PS When we perform manual Full GC no such CPU spikes occured

**Comment by fatum** [comment]
More details:

![screen shot 2017-02-28 at 16 02 16](https://cloud.githubusercontent.com/assets/313161/23406313/616e9e72-fdcf-11e6-8eaf-1accc4693eb4.png)
![screen shot 2017-02-28 at 16 02 00](https://cloud.githubusercontent.com/assets/313161/23406317/64eff78a-fdcf-11e6-94ec-88934b47de59.png)
![screen shot 2017-02-28 at 16 01 50](https://cloud.githubusercontent.com/assets/313161/23406325/6ff079de-fdcf-11e6-8ba9-587b114258cb.png)

**Comment by fatum** [comment]
Fixed in jruby 9.1.8

https://github.com/jruby/jruby/issues/3612#issuecomment-286424340
